x-pack/plugins/enterprise_search/server/collectors/app_search/telemetry.ts

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { set } from 'lodash';
+import { ISavedObjectsRepository, SavedObjectsServiceStart } from 'src/core/server';
+import { UsageCollectionSetup } from 'src/plugins/usage_collection/server';
+
+import { AS_TELEMETRY_NAME, ITelemetrySavedObject } from '../../saved_objects/app_search/telemetry';
+
+/**
+ * Register the telemetry collector
+ */
+
+interface Dependencies {
+  savedObjects: SavedObjectsServiceStart;
+  usageCollection: UsageCollectionSetup;
+}
+
+export const registerTelemetryUsageCollector = ({
+  usageCollection,
+  savedObjects,
+}: Dependencies) => {
+  const upgradeAssistantUsageCollector = usageCollection.makeUsageCollector({
+    type: AS_TELEMETRY_NAME,
+    fetch: async () => fetchTelemetryMetrics(savedObjects),
+  });
+  usageCollection.registerCollector(upgradeAssistantUsageCollector);
+};
+
+/**
+ * Fetch the aggregated telemetry metrics from our saved objects
+ */
+
+const fetchTelemetryMetrics = async (savedObjects: SavedObjectsServiceStart) => {
+  const savedObjectsRepository = savedObjects.createInternalRepository();
+  const savedObjectAttributes = await getSavedObjectAttributesFromRepo(savedObjectsRepository);
+
+  const defaultTelemetrySavedObject = {
+    ui_viewed: {
+      setup_guide: 0,
+      engines_overview: 0,
+    },
+    ui_error: {
+      cannot_connect: 0,
+      no_as_account: 0,
+    },
+    ui_clicked: {
+      header_launch_button: 0,
+      engine_table_link: 0,
+    },
+  };
+
+  // If we don't have an existing/saved telemetry object, return the default
+  if (!savedObjectAttributes) {
+    return defaultTelemetrySavedObject;
+  }
+
+  // Iterate through each attribute key and set its saved values
+  const attributeKeys = Object.keys(savedObjectAttributes);
+  const telemetryObj = defaultTelemetrySavedObject;
+  attributeKeys.forEach((key: string) => {
+    set(telemetryObj, key, savedObjectAttributes[key]);
+  });
+
+  return telemetryObj as ITelemetrySavedObject;
```

Comment:
Sure!

BTW, not sure if this is just me, but I haven't actually had Kibana itself throw any errors when my type definitions are incorrect locally (or even non-existent). Is this something that we should be relying on IDEs for and not the actual Kibana compiler?

Reply:
Fixed: https://github.com/constancecchen/kibana/pull/5/files#diff-66697ea3a0438ad7117559760f4d9550R41

Reply:
The only thing they mention in their [Contributing](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#linting) guide is that they run TSLint on a pre-commit hook. But I don't think that would catch Type errors would it?

Reply:
Yeah I'm pretty baffled. Definitely let me know if you figure this out! 😬

Reply:
Their [TYPESCRIPT.md](https://github.com/elastic/kibana/blob/master/TYPESCRIPT.md) docs also don't have much info in that regard...

My only thought is maybe they haven't turned on strict type-checking yet since they're still in the progress of migrating fully to Typescript perhaps?